File: tests/e2e_test.py
import subprocess
import os
import pytest

# Test cases

def test_copy_file_success(tmp_path):
    # Create a dummy source file
    source_file = tmp_path / "source.txt"
    source_file.write_text("This is the source file.")

    # Define the destination file
    dest_file = tmp_path / "destination.txt"

    # Run the CLI command
    result = subprocess.run(["cargo", "run", "--", str(source_file), str(dest_file)], capture_output=True, text=True)

    # Assert the command was successful
    assert result.returncode == 0
    assert dest_file.read_text() == "This is the source file."


def test_copy_file_overwrite(tmp_path):
    # Create a dummy source file
    source_file = tmp_path / "source.txt"
    source_file.write_text("This is the source file.")

    # Create a destination file with some content
    dest_file = tmp_path / "destination.txt"
    dest_file.write_text("This is the original destination file content.")

    # Run the CLI command
    result = subprocess.run(["cargo", "run", "--", str(source_file), str(dest_file)], capture_output=True, text=True)

    # Assert the command was successful and the destination file was overwritten
    assert result.returncode == 0
    assert dest_file.read_text() == "This is the source file."


def test_copy_file_source_not_found(tmp_path):
    # Define the source and destination files
    source_file = tmp_path / "nonexistent_source.txt"
    dest_file = tmp_path / "destination.txt"

    # Run the CLI command
    result = subprocess.run(["cargo", "run", "--", str(source_file), str(dest_file)], capture_output=True, text=True)

    # Assert the command failed with a non-zero exit code
    assert result.returncode != 0
    assert "No such file or directory" in result.stderr


def test_copy_file_invalid_destination(tmp_path):
    # Create a dummy source file
    source_file = tmp_path / "source.txt"
    source_file.write_text("This is the source file.")

    # Define an invalid destination path
    dest_file = tmp_path / "invalid_dir" / "destination.txt"

    # Run the CLI command
    result = subprocess.run(["cargo", "run", "--", str(source_file), str(dest_file)], capture_output=True, text=True)

    # Assert the command failed with a non-zero exit code
    assert result.returncode != 0
    assert "No such file or directory" in result.stderr

File: tests/e2e_test.rs
use std::process::Command;
use std::fs;
use std::path::Path;

#[test]
fn test_file_copy()
{
    // Create a dummy source file
    let source_file = "test_source.txt";
    let destination_file = "test_destination.txt";
    let content = "This is a test file.";
    fs::write(source_file, content).expect("Unable to write to source file");

    // Run the file_copier CLI
    let status = Command::new("cargo")
        .args(&["run", "--", source_file, destination_file])
        .status()
        .expect("Failed to execute command");

    // Assert that the command was successful
    assert!(status.success());

    // Verify that the destination file exists
    assert!(Path::new(destination_file).exists());

    // Verify that the destination file contains the same content as the source file
    let destination_content = fs::read_to_string(destination_file).expect("Unable to read destination file");
    assert_eq!(content, destination_content);

    // Clean up the created files
    fs::remove_file(source_file).expect("Unable to remove source file");
    fs::remove_file(destination_file).expect("Unable to remove destination file");
}
